crypto/rsa_extra/rsassa_pss_asn1.c

```diff
@@ -0,0 +1,317 @@
+    9,
+};
+
+// Used to check if the oid is in the supported five hash functions.
+// Section 2.1. https://tools.ietf.org/html/rfc4055#page-4
+static const RSA_PSS_SUPPORTED_ALGOR *const rsa_pss_hash_functions[] = {
+    &sha1_func, &sha224_func, &sha256_func, &sha384_func, &sha512_func,
+};
+
+const RSA_PSS_SUPPORTED_ALGOR MGF1 = {
+    NID_mgf1,
+    // 1.2.840.113549.1.1.8
+    {0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x08},
+    9,
+};
+
+// Used to check if the oid is in the supported mask generation functions.
+// Section 2.2. https://tools.ietf.org/html/rfc4055#page-4
+static const RSA_PSS_SUPPORTED_ALGOR *const rsa_pss_mg_functions[] = {
+    &MGF1,
+};
+
+// parse_oid return one on success and zero on failure.
+// Given |oid| and |supported_algors|, if match, set |*out| to be the algorithm
+// identifier.
+static int parse_oid(CBS *oid,
+                     const RSA_PSS_SUPPORTED_ALGOR *const supported_algors[],
+                     size_t size, RSA_ALGOR_IDENTIFIER **out) {
+  for (size_t i = 0; i < size; i++) {
+    const RSA_PSS_SUPPORTED_ALGOR *supported_algr = supported_algors[i];
+    if (CBS_len(oid) == supported_algr->oid_len &&
+        OPENSSL_memcmp(CBS_data(oid), supported_algr->oid,
+                       supported_algr->oid_len) == 0) {
+      *out = RSA_ALGOR_IDENTIFIER_new();
+      if (*out != NULL) {
+        (*out)->nid = supported_algr->nid;
+      }
+      return (*out) != NULL;
+    }
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_UNSUPPORTED_ALGORITHM);
+  return 0;
+}
+
+// For One-way Hash Functions:
+// All implementations MUST accept both NULL and absent parameters as legal and
+// equivalent encodings. See 2.1. https://tools.ietf.org/html/rfc4055#page-5
+static int is_absent_or_null(CBS *params) {
+  CBS null;
+  return (CBS_len(params) == 0) ||
+         (CBS_get_asn1(params, &null, CBS_ASN1_NULL) && (CBS_len(&null) == 0) &&
+          (CBS_len(params) == 0));
+}
+
+// Decode One-way Hash Functions: return one on success and zero on failure.
+// See 2.1. https://tools.ietf.org/html/rfc4055#page-5
+static int decode_one_way_hash(CBS *cbs, RSA_ALGOR_IDENTIFIER **hash_algor) {
+  CBS seq, oid;
+  if (CBS_get_asn1(cbs, &seq, CBS_ASN1_SEQUENCE) && (CBS_len(cbs) == 0) &&
+      CBS_get_asn1(&seq, &oid, CBS_ASN1_OBJECT) && is_absent_or_null(&seq) &&
+      parse_oid(&oid, rsa_pss_hash_functions,
+                OPENSSL_ARRAY_SIZE(rsa_pss_hash_functions), hash_algor)) {
+    return 1;
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_DECODE_ERROR);
+  return 0;
+}
+
+// Decode Mask Generation Functions: return one on success and zero on failure.
+// See 2.2. https://tools.ietf.org/html/rfc4055#page-5
+static int decode_mask_gen(CBS *cbs, RSA_MGA_IDENTIFIER **mga) {
+  CBS seq, mgf1_oid, hash_seq, hash_oid;
+  RSA_ALGOR_IDENTIFIER *mgf1 = NULL;
+  RSA_ALGOR_IDENTIFIER *hash_algor = NULL;
+  if (CBS_get_asn1(cbs, &seq, CBS_ASN1_SEQUENCE) && (CBS_len(cbs) == 0) &&
+      CBS_get_asn1(&seq, &mgf1_oid, CBS_ASN1_OBJECT) &&
```

Comment:
Should we bail early here if `mfg1` is not in the list of supported mask generation algorithm?

Reply:
Done.